reco_utils/recommender/deeprec/DataModel/ImplicitCF.py

```diff
@@ -0,0 +1,217 @@
+# Licensed under the MIT License.
+
+import random
+import numpy as np
+import pandas as pd
+import scipy.sparse as sp
+import time
+from reco_utils.common.constants import (
+    DEFAULT_ITEM_COL,
+    DEFAULT_USER_COL,
+    DEFAULT_RATING_COL,
+    DEFAULT_PREDICTION_COL,
+)
+
+
+class ImplicitCF(object):
+    """Data processing class for GCN models which use implicit feedback.
+
+    Initialize train and test set, create normalized adjacency matrix and sample data for training epochs.
+    
+    """
+
+    def __init__(
+        self,
+        train,
+        test=None,
+        adj_dir=None,
+        col_user=DEFAULT_USER_COL,
+        col_item=DEFAULT_ITEM_COL,
+        col_rating=DEFAULT_RATING_COL,
+        col_prediction=DEFAULT_PREDICTION_COL,
+        seed=None,
+    ):
+        """Constructor 
+        
+        Args:
+            adj_dir (str): Directory to save / load adjacency matrices. If it is None, adjacency
+                matrices will be created and will not be saved.
+            train (pd.DataFrame): Training data with at least columns (col_user, col_item, col_rating).
+            test (pd.DataFrame): Test data with at least columns (col_user, col_item, col_rating).
+                test can be None, if so, we only process the training data.
+            col_user (str): User column name.
+            col_item (str): Item column name.
+            col_rating (str): Rating column name. 
+            seed (int): Seed.
+        
+        """
+        self.user_idx = None
+        self.item_idx = None
+        self.adj_dir = adj_dir
+        self.col_user = col_user
+        self.col_item = col_item
+        self.col_rating = col_rating
+        self.col_prediction = col_prediction
+        self.train, self.test = self._data_processing(train, test)
+        self._init_train_data()
+
+        random.seed(seed)
+
+    def _data_processing(self, train, test):
+        """Process the dataset to reindex userID and itemID and only keep records with ratings greater than 0.
+
+        Args:
+            train (pd.DataFrame): Training data with at least columns (col_user, col_item, col_rating). 
+            test (pd.DataFrame): Test data with at least columns (col_user, col_item, col_rating).
+                test can be None, if so, we only process the training data.
+
+        Returns:
+            list: train and test pd.DataFrame Dataset, which have been reindexed and filtered.
+        
+        """
+        df = train if test is None else train.append(test)
+
+        if self.user_idx is None:
+            user_idx = df[[self.col_user]].drop_duplicates().reindex()
+            user_idx[self.col_user + "_idx"] = np.arange(len(user_idx))
+            self.n_users = len(user_idx)
+            self.user_idx = user_idx
+
+            self.user2id = dict(
+                zip(user_idx[self.col_user], user_idx[self.col_user + "_idx"])
+            )
+            self.id2user = {self.user2id[k]: k for k in self.user2id}
```

Comment:
Reverse table can be simply
`dict(zip(user_idx[self.col_user + "_idx"], user_idx[self.col_user]))`.
This will be faster as well.

Reply:
done